benches/backtracking.rs

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This bench monitors the performance of backtracking and term intersection.
+//!
+//! Dependencies are constructed in a way that all versions need to be tested before
+//! pubgrub can determine that no solution exists.
+
+use criterion::*;
+use pubgrub::OfflineDependencyProvider;
+use version_ranges::Ranges;
+
+fn bench(c: &mut Criterion, case: &str, package_count: u32, version_count: u32) {
+    let mut dependency_provider = OfflineDependencyProvider::<u32, Ranges<u32>>::new();
+    let r = Ranges::strictly_lower_than(version_count);
+    dependency_provider.add_dependencies(0u32, 0u32, [(1u32, r)]);
+
+    for n in 1..package_count {
+        for v in 0..version_count {
+            let r = Ranges::strictly_lower_than(v);
+            dependency_provider.add_dependencies(n, v, [(n + 1, r)]);
+        }
+    }
+
+    c.bench_function(case, |b| {
+        b.iter(|| {
+            let _ = pubgrub::resolve(&dependency_provider, 0u32, 0u32);
+        })
+    });
+}
+
+fn bench_group(c: &mut Criterion) {
+    bench(c, "backtracking_small", 5, 60);
+    bench(c, "backtracking_medium", 5, 200);
+    bench(c, "backtracking_large", 5, 500);
```

Comment:
Why 60, 200, 500? Why only 5 layers?

Reply:
These values are not special, there were adjusted so that the benchmark takes around 10 times more time between `backtracking_medium` and `backtracking_small` and between `backtracking_large` and `backtracking_medium` (here the resolve time is cubic over the number of versions).

The only constraint is that I want one benchmark with less than 63 versions and another with more than 63 versions, so that I can compare performance of u64 version set in future commits.

Reply:
I have had a few minutes to play with it locally. Running the `500` takes 16 min for me witch seams like a long time to be waiting for results.

> (here the resolve time is cubic over the number of versions).

I used [`benchmark_group`](https://docs.rs/criterion/latest/criterion/struct.BenchmarkGroup.html) and `bench_with_input` to make a nice graph.

![lines](https://github.com/user-attachments/assets/17aa5fdc-c1be-4be9-b3e5-b6c02e2aba2f)

Which is definitely not linear.  Throwing it in Excel a quadratic fit the data quite well `R=0.995` (with the intercept pin to zero). But 5 points  may not provide enough data to distinguish  between a quadratic in a cubic.

Reply:
By the way, this micro benchmark gets a lot faster when using #257. 

```
backtracking/60         time:   [2.4683 ms 2.4746 ms 2.4825 ms]
                        change: [-91.340% -91.314% -91.284%] (p = 0.00 < 0.05)
                        Performance has improved.
backtracking/100        time:   [6.5867 ms 6.5943 ms 6.6021 ms]
                        change: [-94.393% -94.384% -94.376%] (p = 0.00 < 0.05)
                        Performance has improved.
backtracking/150        time:   [14.602 ms 14.617 ms 14.632 ms]
                        change: [-95.852% -95.847% -95.843%] (p = 0.00 < 0.05)
                        Performance has improved.
backtracking/200        time:   [25.453 ms 25.489 ms 25.525 ms]
                        change: [-96.715% -96.710% -96.705%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Reply:
I committed two new benchmarks which will not be impacted if #257 is merged.